zOS: Add 31/64-bit interoperability staging

joransiu opened the pull request:
- Copy the customized JNI header files to include31 directory to JDK_IMAGE base directory.
- Copy the libjvm31.so library and libjvm31.x sidedeck to server / j9vm.

Signed-off-by: Joran Siu <joransiu@ca.ibm.com>

joransiu commented:
This change depends on : https://github.com/eclipse-openj9/openj9/pull/13365 to have libjvm31.x and libjvm31.so being built in CMake enabled builds.  It has some additional JNI header and side-deck related changes that is in `openj9-openjdk-jdk11` extensions repo, which is not in this one -- I pulled them along to keep things consistent.   

Once this PR is reviewed, I'll backport to `openj9-openjdk-jdk11` and `openj9-openjdk-jdk17`.

keithc-ca commented on closed/custom/Images-pre.gmk:
```diff
@@ -18,6 +18,20 @@
 # 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
 # ===========================================================================
 
+ifeq ($(OPENJDK_TARGET_OS), zos)
+	# Customized z/OS 31/64-bit Interoperability JNI header files.
+	# These files are generated as part of OpenJ9 build.
+	INCLUDE31_SOURCE_DIR := $(OPENJ9_VM_BUILD_DIR)/include31
+
+	# Stage into JDK base image 'include31' subdirectory
+	$(eval $(call SetupCopyFiles, JDK_COPY_INCLUDE31_HEADERS, \
+		SRC := $(INCLUDE31_SOURCE_DIR), \
+		FILES := $(call FindFiles, $(INCLUDE31_SOURCE_DIR)/), \
+		DEST := $(JDK_IMAGE_DIR)/include31/, \
+	))
+	JDK_TARGETS += $(JDK_COPY_INCLUDE31_HEADERS)
+endif
```
These lines cannot begin with tabs - please use spaces.

Please remove the trailing slashes in the call to `FindFiles` and in the value of `DEST`, and add a comment on the `endif` line.

keithc-ca commented on closed/custom/modules/java.base/Copy.gmk:
```diff
@@ -30,6 +30,28 @@ $(call openj9_copy_files,, \
+$(OPENJ9_TOPDIR)/runtime/include/jni_convert.h \
+$(INCLUDE_TARGET_DIR)/jni_convert.h)
+
+# Customized z/OS 31/64-bit Interoperability JNI header files.
+# These files are generated as part of OpenJ9 build.
```
Why is 'Interoperability' capitalized?
Please add the missing 'an' in 'part of an OpenJ9 build'.

keithc-ca commented on closed/custom/modules/java.base/Copy.gmk:
```diff
@@ -30,6 +30,28 @@ $(call openj9_copy_files,, \
+# Customized z/OS 31/64-bit Interoperability JNI header files.
+# These files are generated as part of OpenJ9 build.
+INCLUDE31_SOURCE_DIR := $(OPENJ9_VM_BUILD_DIR)/include31
+INCLUDE31_TARGET_DIR := $(INCLUDE_TARGET_DIR)/../include31
```
Have you verified that using the name `include31` works with `jlink`?

joransiu replied:
Thanks @keithc-ca.  This is a good point -- I hadn't realized jlink images include the headers.  I'll confirm.

keithc-ca replied:
It's actually the `.jmod` files we are concerned about; they are inputs to `jlink` for producing (possibly custom) images. Some paths are treated specially in `.jmod` files (e.g. `classes/*`); we just need to play by the rules.

joransiu replied:
Right. I see the following in the `java.base.jmod`, not immediately sure how to get additional directories to be included.

```sh
> jar tf java.base.jmod | grep include                                                                                                                                                                     
include/jni.h
include/jvmticmlr.h
include/zos/jni_md.h
include/ibmjvmti.h
include/jvmti.h
include/jni_convert.h
include/classfile_constants.h
```

joransiu replied:
Looks like it's related to this: https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/make/CreateJmods.gmk#L52-L53  though not sure how `include` itself is specified for java.base.

joransiu replied:
Looks like the module specific header files are staged into: ` $(SUPPORT_OUTPUTDIR)/modules_include/$(MODULE)` based on:

https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/make/common/modules/CopyCommon.gmk#L36-L63

joransiu replied:
From what I can tell, jmod appears to support only `--header-files` and will stage files in the given directory into `include` directory of the output jmod file.  Looks like we'd have to modify jmod itself if we want a parallel `include31` directory.  Alternatively, we can have `include31` as a subdirectory within `include` folder by staging the files into modules_include target -- and just document the updated location in migration docs.   There will be another change  to compile /link commands of JNI natives for the libjvm31.x sidedeck file from Java 8, due to lib/s390x/j9vm -> lib/j9vm.   @keithc-ca @pshipton , thoughts?

keithc-ca replied:
I think it would be simpler to put the new header files in a new subdirectory of `include`, e.g. `include/jni31` or similar. Applications that need to use the 31-bit shim API would compile with `{sdk}/include/jni31` on the include path (instead of `{sdk}/include31`).

keithc-ca replied:
BTW, a `.jmod` file is not really in in `jar` format; you should use `jmod`:
```
$ ./bin/jmod list jmods/java.base.jmod | grep include | sort
include/classfile_constants.h
include/ibmjvmti.h
include/jni.h
include/jvmticmlr.h
include/jvmti.h
include/linux/jni_md.h
```

joransiu replied:
Thanks @keithc-ca.  I got some initial feedback from 31/64-bit users... no one is against changing it to subdirectory of `include/` for Java 11+.  In some ways, `include/jni31` as you suggested is cleaner.   This simplifies this PR, as none of the `Image-pre.gmk` changes are needed.  I'll touch base with some other users who will be back from Thanksgiving next week to see if anyone raises additional concerns.

joransiu replied:
Got feedback from users that the suggested change to include/jni31 was acceptable.  I'll update this PR with the necessary changes, and also address the other review comments.  Thanks.

keithc-ca commented on closed/custom/modules/java.base/Copy.gmk:
```diff
@@ -30,6 +30,28 @@ $(call openj9_copy_files,, \
 
+ifeq ($(OPENJDK_TARGET_OS), zos)
+$(call openj9_copy_files,, \
+$(OPENJ9_TOPDIR)/runtime/include/jni_convert.h \
+$(INCLUDE_TARGET_DIR)/jni_convert.h)
```
Please indent lines 35, 36.

keithc-ca commented on closed/custom/modules/java.base/Copy.gmk:
```diff
@@ -84,6 +106,26 @@ $(call openj9_copy_files_and_debuginfos, \
+	$(addsuffix libjvm31.x, \
+		$(OPENJ9_VM_BUILD_DIR)/lib/ \
+		$(addprefix $(LIB_DST_DIR), /j9vm/ /server/)))
 endif
```
This should be commented:
```
endif # zos
```

joransiu commented:
@keithc-ca : I've removed all the Image-Pre.gmk changes, and updated the target directory of the 31-bit JNI headers to include/jni31 of java.base in the latest commit.

keithc-ca commented on closed/custom/modules/java.base/Copy.gmk:
```diff
@@ -30,6 +30,28 @@ $(call openj9_copy_files,, \
 	$(OPENJ9_VM_BUILD_DIR)/include/jvmti.h \
 	$(INCLUDE_TARGET_DIR)/jvmti.h)
 
+ifeq ($(OPENJDK_TARGET_OS), zos)
```
nit: for consistency, the arguments should be reversed (see [OpenJ9.gmk](https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/closed/OpenJ9.gmk#L94)):
```
ifeq (zos,$(OPENJDK_TARGET_OS))
```
and the endif line should just be:
```
endif # zos
```
Same goes for the conditional block which starts on line 102.

joransiu replied:
Updated under 48c8f92. Thanks!

keithc-ca commented:
Please repeat the second part of ibmruntimes/openj9-openjdk-jdk11#459 to reduce differences between versions.

keithc-ca commented:
Jenkins compile alinux64 jdknext

keithc-ca commented:
@pshipton I'll let you merge this when you're satisfied with the changes.
